manifest/ocischema/index.go

```diff
@@ -0,0 +1,174 @@
+		}
+
+		if m.MediaType != "" && m.MediaType != v1.MediaTypeImageIndex {
+			err = fmt.Errorf("if present, mediaType in image index should be '%s' not '%s'",
+				v1.MediaTypeImageIndex, m.MediaType)
+
+			return nil, distribution.Descriptor{}, err
+		}
+
+		dgst := digest.FromBytes(b)
+		return m, distribution.Descriptor{Digest: dgst, Size: int64(len(b)), MediaType: v1.MediaTypeImageIndex}, err
+	}
+	err := distribution.RegisterManifestSchema(v1.MediaTypeImageIndex, imageIndexFunc)
+	if err != nil {
+		panic(fmt.Sprintf("Unable to register OCI Image Index: %s", err))
+	}
+}
+
+// A ManifestDescriptor references a platform-specific manifest.
+type ManifestDescriptor struct {
+	distribution.Descriptor
+
+	// Platform specifies which platform the manifest pointed to by the
+	// descriptor runs on.
+	Platform *v1.Platform `json:"platform,omitempty"`
+}
+
+// ImageIndex references manifests for various platforms.
+type ImageIndex struct {
+	manifest.Versioned
+
+	// Manifests references a list of manifests
+	Manifests []ManifestDescriptor `json:"manifests"`
+
+	// Annotations is an optional field that contains arbitrary metadata for the
+	// image index
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// References returns the distribution descriptors for the referenced image
+// manifests.
+func (ii ImageIndex) References() []distribution.Descriptor {
+	dependencies := make([]distribution.Descriptor, len(ii.Manifests))
+	for i := range ii.Manifests {
+		dependencies[i] = ii.Manifests[i].Descriptor
+		dependencies[i].Platform = ii.Manifests[i].Platform
+	}
+
+	return dependencies
+}
+
+// DeserializedImageIndex wraps ManifestList with a copy of the original
+// JSON.
+type DeserializedImageIndex struct {
+	ImageIndex
+
+	// canonical is the canonical byte representation of the Manifest.
+	canonical []byte
+}
+
+// FromDescriptors takes a slice of descriptors and a map of annotations, and
+// returns a DeserializedManifestList which contains the resulting manifest list
+// and its JSON representation. If annotations is nil or empty then the
+// annotations property will be omitted from the JSON representation.
+func FromDescriptors(descriptors []ManifestDescriptor, annotations map[string]string) (*DeserializedImageIndex, error) {
+	return fromDescriptorsWithMediaType(descriptors, annotations, v1.MediaTypeImageIndex)
+}
+
+// fromDescriptorsWithMediaType is for testing purposes, it's useful to be able to specify the media type explicitly
+func fromDescriptorsWithMediaType(descriptors []ManifestDescriptor, annotations map[string]string, mediaType string) (*DeserializedImageIndex, error) {
+	m := ImageIndex{
+		Versioned: manifest.Versioned{
+			SchemaVersion: 2,
+			MediaType:     mediaType,
+		},
+		Annotations: annotations,
+	}
+
+	m.Manifests = make([]ManifestDescriptor, len(descriptors))
+	copy(m.Manifests, descriptors)
+
+	deserialized := DeserializedImageIndex{
+		ImageIndex: m,
+	}
+
+	var err error
```

Comment:
instead of this we could also initialize in the `func` declaration returned values `(d *DeserializedImageIndex, err error)`. We would have to change `deserialized := DeserializedImageIndex{` to `d = DeserializedImageIndex{` and there would be no need for this `err` variable, etc.

Reply:
I avoid named return values in all but functions for the small (a few lines) algorithms. ref: https://go.dev/tour/basics/7

Though I'm willing to use `(_ *DeserializedImageIndex, err error)` because err is so ubiquitous and because this prevents use of "naked returns" which can lead to confusing defects.